periodic-log-dump-stats: feature added

GeorgeA92 opened the pull request:
Aimed to solve https://github.com/scrapy/scrapy/issues/2173

The main idea of this approach is to.. get  **all**  `int` and `float` values of `crawler.stats` (and key names of course) and log.. it's changes (delta) on periodic basis.
on default log interval (60 seconds) it will log numeric stats changes for last minute.

<details><summary>script.py (books to scrape LOG_LEVEL -> INFO)</summary>

```python
import scrapy
from scrapy.crawler import CrawlerProcess

class BooksToScrapeSpider(scrapy.Spider):
    name = "books"
    custom_settings = {
        "DOWNLOAD_DELAY": 0.3,
        "LOG_LEVEL": 'INFO'
    }
    def start_requests(self):
        yield scrapy.Request(url='https://books.toscrape.com/', callback=self.parse)

    def parse(self, response):
        for book_page_link in response.css('.product_pod a::attr(href)').getall():
            yield scrapy.Request(response.urljoin(book_page_link), self.parse_book)
        if next_page := response.css("li.next a::attr(href)").get():
            yield scrapy.Request(url=response.urljoin(next_page), callback=self.parse)

    def parse_book(self, response):
        yield {'response.url': response.url}

if __name__ == "__main__":
    process = CrawlerProcess()
    process.crawl(BooksToScrapeSpider)
    process.start()
```
</details>

<details><summary>log output</summary>

```
2023-02-09 01:27:23 [scrapy.utils.log] INFO: Scrapy 2.8.0 started (bot: scrapybot)
2023-02-09 01:27:23 [scrapy.utils.log] INFO: Versions: lxml 4.9.1.0, libxml2 2.9.14, cssselect 1.1.0, parsel 1.6.0, w3lib 1.21.0, Twisted 22.2.0, Python 3.10.8 | packaged by conda-forge | (main, Nov 24 2022, 14:07:00) [MSC v.1916 64 bit (AMD64)], pyOpenSSL 22.0.0 (OpenSSL 1.1.1s  1 Nov 2022), cryptography 37.0.4, Platform Windows-10-10.0.22621-SP0
2023-02-09 01:27:23 [scrapy.crawler] INFO: Overridden settings:
{'DOWNLOAD_DELAY': 0.3, 'LOG_LEVEL': 'INFO'}
2023-02-09 01:27:23 [py.warnings] WARNING: <redacted> ScrapyDeprecationWarning: '2.6' is a deprecated value for the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting.

It is also the default value. In other words, it is normal to get this warning if you have not defined a value for the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting. This is so for backward compatibility reasons, but it will change in a future version of Scrapy.

See the documentation of the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting for information on how to handle this deprecation.
  return cls(crawler)

2023-02-09 01:27:23 [scrapy.extensions.telnet] INFO: Telnet Password: d8fd4beeb7768b38
2023-02-09 01:27:23 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.logstats.LogStats']
2023-02-09 01:27:23 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2023-02-09 01:27:23 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.spidermiddlewares.referer.RefererMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
2023-02-09 01:27:23 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2023-02-09 01:27:23 [scrapy.core.engine] INFO: Spider opened
2023-02-09 01:27:24 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2023-02-09 01:27:24 [scrapy.extensions.logstats] INFO: Dumping periodic (60.0 seconds) Scrapy stats (0):
{'log_count/INFO': 8, 'log_count/WARNING': 1}
2023-02-09 01:27:24 [scrapy.extensions.telnet] INFO: Telnet console listening on 127.0.0.1:6024
2023-02-09 01:28:24 [scrapy.extensions.logstats] INFO: Crawled 161 pages (at 161 pages/min), scraped 151 items (at 151 items/min)
2023-02-09 01:28:24 [scrapy.extensions.logstats] INFO: Dumping periodic (60.0 seconds) Scrapy stats (1):
{'downloader/request_bytes': 60517,
 'downloader/request_count': 177,
 'downloader/request_method_count/GET': 177,
 'downloader/response_bytes': 640456,
 'downloader/response_count': 161,
 'downloader/response_status_count/200': 161,
 'dupefilter/filtered': 200,
 'httpcompression/response_bytes': 3430779,
 'httpcompression/response_count': 161,
 'item_scraped_count': 151,
 'log_count/INFO': 3,
 'log_count/WARNING': 0,
 'request_depth_max': 10,
 'response_received_count': 161,
 'scheduler/dequeued': 177,
 'scheduler/dequeued/memory': 177,
 'scheduler/enqueued': 211,
 'scheduler/enqueued/memory': 211}
2023-02-09 01:29:24 [scrapy.extensions.logstats] INFO: Crawled 324 pages (at 163 pages/min), scraped 303 items (at 152 items/min)
2023-02-09 01:29:24 [scrapy.extensions.logstats] INFO: Dumping periodic (60.0 seconds) Scrapy stats (2):
{'downloader/request_bytes': 55640,
 'downloader/request_count': 163,
 'downloader/request_method_count/GET': 163,
 'downloader/response_bytes': 645464,
 'downloader/response_count': 163,
 'downloader/response_status_count/200': 163,
 'dupefilter/filtered': 200,
 'httpcompression/response_bytes': 3479739,
 'httpcompression/response_count': 163,
 'item_scraped_count': 152,
 'log_count/INFO': 2,
 'log_count/WARNING': 0,
 'request_depth_max': 10,
 'response_received_count': 163,
 'scheduler/dequeued': 163,
 'scheduler/dequeued/memory': 163,
 'scheduler/enqueued': 210,
 'scheduler/enqueued/memory': 210}
2023-02-09 01:30:24 [scrapy.extensions.logstats] INFO: Crawled 487 pages (at 163 pages/min), scraped 457 items (at 154 items/min)
2023-02-09 01:30:24 [scrapy.extensions.logstats] INFO: Dumping periodic (60.0 seconds) Scrapy stats (3):
{'downloader/request_bytes': 56059,
 'downloader/request_count': 163,
 'downloader/request_method_count/GET': 163,
 'downloader/response_bytes': 660927,
 'downloader/response_count': 163,
 'downloader/response_status_count/200': 163,
 'dupefilter/filtered': 200,
 'httpcompression/response_bytes': 3536010,
 'httpcompression/response_count': 163,
 'item_scraped_count': 154,
 'log_count/INFO': 2,
 'log_count/WARNING': 0,
 'request_depth_max': 10,
 'response_received_count': 163,
 'scheduler/dequeued': 163,
 'scheduler/dequeued/memory': 163,
 'scheduler/enqueued': 210,
 'scheduler/enqueued/memory': 210}
2023-02-09 01:31:24 [scrapy.extensions.logstats] INFO: Crawled 650 pages (at 163 pages/min), scraped 611 items (at 154 items/min)
2023-02-09 01:31:24 [scrapy.extensions.logstats] INFO: Dumping periodic (60.0 seconds) Scrapy stats (4):
{'downloader/request_bytes': 55443,
 'downloader/request_count': 163,
 'downloader/request_method_count/GET': 163,
 'downloader/response_bytes': 636596,
 'downloader/response_count': 163,
 'downloader/response_status_count/200': 163,
 'dupefilter/filtered': 180,
 'httpcompression/response_bytes': 3426316,
 'httpcompression/response_count': 163,
 'item_scraped_count': 154,
 'log_count/INFO': 2,
 'log_count/WARNING': 0,
 'request_depth_max': 9,
 'response_received_count': 163,
 'scheduler/dequeued': 163,
 'scheduler/dequeued/memory': 163,
 'scheduler/enqueued': 189,
 'scheduler/enqueued/memory': 189}
2023-02-09 01:32:24 [scrapy.extensions.logstats] INFO: Crawled 810 pages (at 160 pages/min), scraped 762 items (at 151 items/min)
2023-02-09 01:32:24 [scrapy.extensions.logstats] INFO: Dumping periodic (60.0 seconds) Scrapy stats (5):
{'downloader/request_bytes': 53771,
 'downloader/request_count': 160,
 'downloader/request_method_count/GET': 160,
 'downloader/response_bytes': 620190,
 'downloader/response_count': 160,
 'downloader/response_status_count/200': 160,
 'dupefilter/filtered': 180,
 'httpcompression/response_bytes': 3345432,
 'httpcompression/response_count': 160,
 'item_scraped_count': 151,
 'log_count/INFO': 2,
 'log_count/WARNING': 0,
 'request_depth_max': 9,
 'response_received_count': 160,
 'scheduler/dequeued': 160,
 'scheduler/dequeued/memory': 160,
 'scheduler/enqueued': 189,
 'scheduler/enqueued/memory': 189}
2023-02-09 01:33:24 [scrapy.extensions.logstats] INFO: Crawled 972 pages (at 162 pages/min), scraped 921 items (at 159 items/min)
2023-02-09 01:33:24 [scrapy.extensions.logstats] INFO: Dumping periodic (60.0 seconds) Scrapy stats (6):
{'downloader/request_bytes': 55483,
 'downloader/request_count': 162,
 'downloader/request_method_count/GET': 162,
 'downloader/response_bytes': 618267,
 'downloader/response_count': 162,
 'downloader/response_status_count/200': 162,
 'dupefilter/filtered': 40,
 'httpcompression/response_bytes': 3170892,
 'httpcompression/response_count': 162,
 'item_scraped_count': 159,
 'log_count/INFO': 2,
 'log_count/WARNING': 0,
 'request_depth_max': 2,
 'response_received_count': 162,
 'scheduler/dequeued': 162,
 'scheduler/dequeued/memory': 162,
 'scheduler/enqueued': 41,
 'scheduler/enqueued/memory': 41}
2023-02-09 01:33:54 [scrapy.core.engine] INFO: Closing spider (finished)
2023-02-09 01:33:54 [scrapy.statscollectors] INFO: Dumping Scrapy stats:
{'downloader/request_bytes': 358595,
 'downloader/request_count': 1050,
 'downloader/request_method_count/GET': 1050,
 'downloader/response_bytes': 4126715,
 'downloader/response_count': 1050,
 'downloader/response_status_count/200': 1050,
 'dupefilter/filtered': 1000,
 'elapsed_time_seconds': 390.017596,
 'finish_reason': 'finished',
 'finish_time': datetime.datetime(2023, 2, 8, 23, 33, 54, 128634),
 'httpcompression/response_bytes': 21921006,
 'httpcompression/response_count': 1050,
 'item_scraped_count': 1000,
 'log_count/INFO': 23,
 'log_count/WARNING': 1,
 'request_depth_max': 50,
 'response_received_count': 1050,
 'scheduler/dequeued': 1050,
 'scheduler/dequeued/memory': 1050,
 'scheduler/enqueued': 1050,
 'scheduler/enqueued/memory': 1050,
 'start_time': datetime.datetime(2023, 2, 8, 23, 27, 24, 111038)}
2023-02-09 01:33:54 [scrapy.core.engine] INFO: Spider closed (finished)

Process finished with exit code 0


```
</details>

wRAR commented:
Should it be configurable?

GeorgeA92 commented:
Configurable (in terms of enable it by default or not). - I think Yes.
On scrapy projects where decided to set `LOG_LEVEL` from default `DEBUG` to `INFO` because of large number of log entries, enabling this by default means significant increasing log size (and in the same time it will definitely may require much more info than items/pages per minute from regular logstats) and probably not expected

At this stage I aim to implement this similar to `DUPEFILTER_DEBUG` in dupefilter (log extended stats for all runtime with explicitly set setting like `LOG_STATS_EXTENDED:True` and by default: log extended stats for first 2-3 iterations, and log message with mention of `LOG_STATS_EXTENDED:True` after ... )

Gallaecio commented on scrapy/extensions/logstats.py:
```diff
@@ -53,6 +56,16 @@ def log(self, spider):
         logger.info(msg, log_args, extra={"spider": spider})
 
+        num_stats = {k: v for k, v in self.stats._stats.items() if isinstance(v, (int, float))}
+        delta = {k: v - self.stats_prev.get(k, 0) for k, v in num_stats.items()}
```
Maybe we could remove 0 values?

Gallaecio commented:
> At this stage I aim to implement this similar to `DUPEFILTER_DEBUG` in dupefilter (log extended stats for all runtime with explicitly set setting like `LOG_STATS_EXTENDED:True` and by default: log extended stats for first 2-3 iterations, and log message with mention of `LOG_STATS_EXTENDED:True` after ... )

I think printing them every time if True and never if False may be best in this case. The dupefilter approach exists because otherwise users may not realize that requests are being dropped. I don’t think that kind of issue exists here, users should know that stats change, and in any case there is no risk of data loss.

Gallaecio reviewed (commented):
I wonder if enabling this extended stat logging should cause the simpler logging message (*Crawled …*) to *not* be logged, since the information is already in the stats.

GeorgeA92 commented:
I have following note about various.. configurability aspects ot this. Each point of this is discussable and feedback for every point required to.. choose direction of this:
1. On this comment https://github.com/scrapy/scrapy/issues/2173#issuecomment-247994786 requested possibility to make output of log entry.. compartible with custom log processors like logstash etc.
2. Implementing p.1. - probably also means that it's required to log this in separate logger object (or even move this functionality other extension)
3. From the other side if output format is customisable, other more verbose(easily readable) output can be applied like:
```
      {'downloader/request_bytes': '413533 (+60517 on 60s),
       'downloader/request_count': '534 (+177 on 60s),
       ..
```
4. Or provide possibility to apply any other user made format by setting, that pointing to user's output format implementation.
5. Also related issues have various requests to make.. customisable.. list of fields returned this is also relevant here because may also affected by any other `int` or `float` stat's params from custom middlewares/extensions. Something like `EXT_STATS_INCLUDE = ['downloader/','scheduler/']` meaning to output only stat params that has specified substring in stat param name.
6. Also it is still a question about.. enabling (or not enabling) this by default.
From my understanding it is required to enable it by default (like notifications from dupefilter, or keeping data from.. the latest.. 5 minutes of runtime prepared to log)
I am sure that on situations like this:
```
[scrapy.extensions.logstats] INFO: Crawled 972 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
[scrapy.extensions.logstats] INFO: Crawled 972 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
```
Changes of stats params can point to root-cause of issues (important/worth enought to enable somehow by default).

Gallaecio commented:
1, 2, 3, 4: I like JSON format more than pprint myself, but I don’t think we need a separate logger or extension, I think a simpler approach would be best, I don’t think we need to make the output customizable, at least not in this initial implementation.

5: I think supporting some kind of filtering makes sense, otherwise things could be too verbose. I am OK with *not* doing it here, but I think it would be nice to have.

6: I personally don’t think it would be wise to enable this by default, assuming by default it would log all stats. Although given we implement a filtering settings, I wonder if we could make that filtering setting limit stats to a limited few that we consider important at run time, including those that the “Crawled N pages…” INFO messages show, and do enable this by default as a replacement for those INFO messages.

GeorgeA92 commented:
Implemented simple filter for list of stat params by settings
can be used like this:
```python
    custom_settings = {
        "DOWNLOAD_DELAY": 0.3,
        "LOG_LEVEL": 'INFO',
        "LOGSTATS_EXTENDED_ENABLED": True,
        # ^ If LOGSTATS_EXTENDED_ENABLED:True extended log stats, output diff of all (int, float) stat params
        # ^ If not set or LOGSTATS_EXTENDED_ENABLED:False - application log regular log stats message instead of ext_stats
        "LOGSTATS_EXTENDED_INCLUDE": ["downloader/", "scheduler/", "log_count/", "item_scraped_count"],
        # ^ output only stat params that countain one of listed substrings, if not set - allow all
        "LOGSTATS_EXTENDED_EXCLUDE": ["scheduler/"]
        # ^ exclude stat params that contain one of listed substrings, if not set - allow all
        # ^ It has precedence over LOGSTATS_EXTENDED_INCLUDE
    }
```
Implemented Json log output for ext_stat using `scrapy.utils.serialize.ScrapyJSONEncoder`. As I aimed to make it compartible with custom logprocessors Json output means that log message should containg only json compatible output (applied suggestion from https://github.com/scrapy/scrapy/pull/5830#pullrequestreview-1299306902 -to  disable regular _Crawled …_ log message if extened log stats enabled

<details><summary>script.py </summary>

```python
import scrapy
from scrapy.crawler import CrawlerProcess

class BooksToScrapeSpider(scrapy.Spider):
    name = "books"
    custom_settings = {
        "DOWNLOAD_DELAY": 0.3,
        "LOG_LEVEL": 'INFO',
        "LOGSTATS_EXTENDED_ENABLED": True,
        # ^ If LOGSTATS_EXTENDED_ENABLED:True extended log stats, output diff of all (int, float) stat params
        # ^ If not set or LOGSTATS_EXTENDED_ENABLED:False - application log regular log stats message instead of ext_stats
        "LOGSTATS_EXTENDED_INCLUDE": ["downloader/", "scheduler/", "log_count/", "item_scraped_count"],
        # ^ output only stat params that countain one of listed substrings, if not set - allow all
        "LOGSTATS_EXTENDED_EXCLUDE": ["scheduler/"]
        # ^ exclude stat params that contain one of listed substrings, if not set - allow all
        # ^ It has precedence over LOGSTATS_EXTENDED_INCLUDE
    }
    def start_requests(self):
        yield scrapy.Request(url='https://books.toscrape.com/', callback=self.parse)

    def parse(self, response):
        for book_page_link in response.css('.product_pod a::attr(href)').getall():
            yield scrapy.Request(response.urljoin(book_page_link), self.parse_book)
        if next_page := response.css("li.next a::attr(href)").get():
            yield scrapy.Request(url=response.urljoin(next_page), callback=self.parse)

    def parse_book(self, response):
        yield {'response.url': response.url}

if __name__ == "__main__":
    process = CrawlerProcess()
    process.crawl(BooksToScrapeSpider)
    process.start()
```
</details>

<details><summary>log_output</summary>

```
2023-03-03 20:17:02 [scrapy.utils.log] INFO: Scrapy 2.8.0 started (bot: scrapybot)
2023-03-03 20:17:02 [scrapy.utils.log] INFO: Versions: lxml 4.9.1.0, libxml2 2.9.14, cssselect 1.1.0, parsel 1.6.0, w3lib 1.21.0, Twisted 22.2.0, Python 3.10.8 | packaged by conda-forge | (main, Nov 24 2022, 14:07:00) [MSC v.1916 64 bit (AMD64)], pyOpenSSL 22.0.0 (OpenSSL 1.1.1s  1 Nov 2022), cryptography 37.0.4, Platform Windows-10-10.0.22621-SP0
2023-03-03 20:17:02 [scrapy.crawler] INFO: Overridden settings:
{'DOWNLOAD_DELAY': 0.3, 'LOG_LEVEL': 'INFO'}
2023-03-03 20:17:02 [py.warnings] WARNING: C:\Users\georg\PycharmProjects\scrapy\scrapy\utils\request.py:232: ScrapyDeprecationWarning: '2.6' is a deprecated value for the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting.

It is also the default value. In other words, it is normal to get this warning if you have not defined a value for the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting. This is so for backward compatibility reasons, but it will change in a future version of Scrapy.

See the documentation of the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting for information on how to handle this deprecation.
  return cls(crawler)

2023-03-03 20:17:02 [scrapy.extensions.telnet] INFO: Telnet Password: ea386d462c9cb7b8
2023-03-03 20:17:02 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.logstats.LogStats']
2023-03-03 20:17:03 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2023-03-03 20:17:03 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.spidermiddlewares.referer.RefererMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
2023-03-03 20:17:03 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2023-03-03 20:17:03 [scrapy.core.engine] INFO: Spider opened
2023-03-03 20:17:03 [scrapy.extensions.logstats] INFO: {
    "log_count/INFO": 7,
    "log_count/WARNING": 1
}
2023-03-03 20:17:03 [scrapy.extensions.telnet] INFO: Telnet console listening on 127.0.0.1:6023
2023-03-03 20:18:03 [scrapy.extensions.logstats] INFO: {
    "downloader/request_bytes": 60577,
    "downloader/request_count": 177,
    "downloader/request_method_count/GET": 177,
    "downloader/response_bytes": 646140,
    "downloader/response_count": 161,
    "downloader/response_status_count/200": 161,
    "item_scraped_count": 150,
    "log_count/INFO": 2,
    "log_count/WARNING": 0
}
2023-03-03 20:19:03 [scrapy.extensions.logstats] INFO: {
    "downloader/request_bytes": 54364,
    "downloader/request_count": 159,
    "downloader/request_method_count/GET": 159,
    "downloader/response_bytes": 628721,
    "downloader/response_count": 159,
    "downloader/response_status_count/200": 159,
    "item_scraped_count": 150,
    "log_count/INFO": 1,
    "log_count/WARNING": 0
}
2023-03-03 20:20:03 [scrapy.extensions.logstats] INFO: {
    "downloader/request_bytes": 54883,
    "downloader/request_count": 160,
    "downloader/request_method_count/GET": 160,
    "downloader/response_bytes": 648285,
    "downloader/response_count": 160,
    "downloader/response_status_count/200": 160,
    "item_scraped_count": 150,
    "log_count/INFO": 1,
    "log_count/WARNING": 0
}
2023-03-03 20:21:03 [scrapy.extensions.logstats] INFO: {
    "downloader/request_bytes": 54827,
    "downloader/request_count": 161,
    "downloader/request_method_count/GET": 161,
    "downloader/response_bytes": 629054,
    "downloader/response_count": 161,
    "downloader/response_status_count/200": 161,
    "item_scraped_count": 151,
    "log_count/INFO": 1,
    "log_count/WARNING": 0
}
2023-03-03 20:22:03 [scrapy.extensions.logstats] INFO: {
    "downloader/request_bytes": 52700,
    "downloader/request_count": 157,
    "downloader/request_method_count/GET": 157,
    "downloader/response_bytes": 608761,
    "downloader/response_count": 157,
    "downloader/response_status_count/200": 157,
    "item_scraped_count": 148,
    "log_count/INFO": 1,
    "log_count/WARNING": 0
}
2023-03-03 20:23:03 [scrapy.extensions.logstats] INFO: {
    "downloader/request_bytes": 56544,
    "downloader/request_count": 165,
    "downloader/request_method_count/GET": 165,
    "downloader/response_bytes": 629688,
    "downloader/response_count": 165,
    "downloader/response_status_count/200": 165,
    "item_scraped_count": 164,
    "log_count/INFO": 1,
    "log_count/WARNING": 0
}
2023-03-03 20:23:35 [scrapy.core.engine] INFO: Closing spider (finished)
2023-03-03 20:23:35 [scrapy.statscollectors] INFO: Dumping Scrapy stats:
{'downloader/request_bytes': 358595,
 'downloader/request_count': 1050,
 'downloader/request_method_count/GET': 1050,
 'downloader/response_bytes': 4127101,
 'downloader/response_count': 1050,
 'downloader/response_status_count/200': 1050,
 'dupefilter/filtered': 1000,
 'elapsed_time_seconds': 392.644847,
 'finish_reason': 'finished',
 'finish_time': datetime.datetime(2023, 3, 3, 18, 23, 35, 796411),
 'httpcompression/response_bytes': 21921006,
 'httpcompression/response_count': 1050,
 'item_scraped_count': 1000,
 'log_count/INFO': 16,
 'log_count/WARNING': 1,
 'request_depth_max': 50,
 'response_received_count': 1050,
 'scheduler/dequeued': 1050,
 'scheduler/dequeued/memory': 1050,
 'scheduler/enqueued': 1050,
 'scheduler/enqueued/memory': 1050,
 'start_time': datetime.datetime(2023, 3, 3, 18, 17, 3, 151564)}
2023-03-03 20:23:35 [scrapy.core.engine] INFO: Spider closed (finished)

Process finished with exit code 0

```
</details>

Remaining issue is.. the gap on end of crawl between the last log stats message and closing spider like on https://github.com/scrapy/scrapy/issues/4110

kmike commented:
An interesting take on this @GeorgeA92! Tbh, I still prefer the most basic approach:

1. Don't use JSON.
2. Log current values, not the delta.

It may look a bit less nice, but it seems it solves the problem, and doesn't have any gotchas. 

JSON can be fine though, no strong opinon on that.

Some possible issues:

1. Some stat values may not be possible to encode to JSON. Not sure how common is that, but I have used such stats in past. If we still support such stats, then we need to ensure not to raise an error when getting a value which can't be put to JSON.
2. The feature is named "logging stats", so we should probably log stats by default; deltas are not stats.
3. Logging deltas makes some assumptions about the stat values: they're numeric, and that it's the change which is interesting. It makes sense for stats where inc_value is used, but not necessarily for stats with .set_value. For example, it looks more valuable to see elapsed_time_seconds as a value, not as delta (it'd be always similar to log stats interval). Other possible stats, where checking the value could be better than checking the delta: current number of filled slots in the downloader, throttle % in scrapy-zyte-api. There are also string values possibile, e.g. those which are used to indicate the current state of the crawler.

I like a lot having an option to log deltas, or having an option to include/exclude values, but I'm not sure if that should be a default (at least unless we change all the option names and the language used to describe the feature).

GeorgeA92 commented:
Added logging crawler stats (at this stage without filtering stat names as on deltas).
And added logging timing info as mentioned on comment above

<details><summary>script.py</summary>

```python
import scrapy
from scrapy.crawler import CrawlerProcess

class BooksToScrapeSpider(scrapy.Spider):
    name = "books"
    custom_settings = \
        {
        "DOWNLOAD_DELAY": 0.3,
        "LOG_LEVEL": 'INFO',
        "LOGSTATS_EXTENDED_ENABLED": True,
        # ^ If LOGSTATS_EXTENDED_ENABLED:True extended log stats, output diff of all (int, float) stat params
        # ^ If not set or LOGSTATS_EXTENDED_ENABLED:False - application log regular log stats message instead if ext_stats
        "LOGSTATS_EXTENDED_INCLUDE": ["downloader/", "scheduler/", "log_count/", "item_scraped_count"],
        # ^ output only stat params that countain one of listed substrings, if not set - allow all
        "LOGSTATS_EXTENDED_EXCLUDE": ["scheduler/"],
        # ^ exclude stat params that contain one of listed substrings, if not set - allow all
        # ^ It has precedence over LOGSTATS_EXTENDED_INCLUDE

        }
    def start_requests(self):
        yield scrapy.Request(url='https://books.toscrape.com/', callback=self.parse)

    def parse(self, response):
        for book_page_link in response.css('.product_pod a::attr(href)').getall():
            yield scrapy.Request(response.urljoin(book_page_link), self.parse_book)
        if next_page := response.css("li.next a::attr(href)").get():
            yield scrapy.Request(url=response.urljoin(next_page), callback=self.parse)

    def parse_book(self, response):
        yield {'response.url': response.url}

if __name__ == "__main__":
    process = CrawlerProcess()
    process.crawl(BooksToScrapeSpider)
    process.start()

```
</details>

<details><summary>log output</summary>

```
2023-03-16 15:36:18 [scrapy.utils.log] INFO: Scrapy 2.8.0 started (bot: scrapybot)
2023-03-16 15:36:18 [scrapy.utils.log] INFO: Versions: lxml 4.9.1.0, libxml2 2.9.14, cssselect 1.1.0, parsel 1.6.0, w3lib 1.21.0, Twisted 22.2.0, Python 3.10.8 | packaged by conda-forge | (main, Nov 24 2022, 14:07:00) [MSC v.1916 64 bit (AMD64)], pyOpenSSL 22.0.0 (OpenSSL 1.1.1s  1 Nov 2022), cryptography 37.0.4, Platform Windows-10-10.0.22621-SP0
2023-03-16 15:36:18 [scrapy.crawler] INFO: Overridden settings:
{'DOWNLOAD_DELAY': 0.3, 'LOG_LEVEL': 'INFO'}
2023-03-16 15:36:18 [py.warnings] WARNING: C:\Users\georg\PycharmProjects\scrapy\scrapy\utils\request.py:232: ScrapyDeprecationWarning: '2.6' is a deprecated value for the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting.

It is also the default value. In other words, it is normal to get this warning if you have not defined a value for the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting. This is so for backward compatibility reasons, but it will change in a future version of Scrapy.

See the documentation of the 'REQUEST_FINGERPRINTER_IMPLEMENTATION' setting for information on how to handle this deprecation.
  return cls(crawler)

2023-03-16 15:36:18 [scrapy.extensions.telnet] INFO: Telnet Password: 1b8c43682f900b1c
2023-03-16 15:36:18 [scrapy.middleware] INFO: Enabled extensions:
['scrapy.extensions.corestats.CoreStats',
 'scrapy.extensions.telnet.TelnetConsole',
 'scrapy.extensions.logstats.LogStats']
2023-03-16 15:36:19 [scrapy.middleware] INFO: Enabled downloader middlewares:
['scrapy.downloadermiddlewares.httpauth.HttpAuthMiddleware',
 'scrapy.downloadermiddlewares.downloadtimeout.DownloadTimeoutMiddleware',
 'scrapy.downloadermiddlewares.defaultheaders.DefaultHeadersMiddleware',
 'scrapy.downloadermiddlewares.useragent.UserAgentMiddleware',
 'scrapy.downloadermiddlewares.retry.RetryMiddleware',
 'scrapy.downloadermiddlewares.redirect.MetaRefreshMiddleware',
 'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware',
 'scrapy.downloadermiddlewares.redirect.RedirectMiddleware',
 'scrapy.downloadermiddlewares.cookies.CookiesMiddleware',
 'scrapy.downloadermiddlewares.httpproxy.HttpProxyMiddleware',
 'scrapy.downloadermiddlewares.stats.DownloaderStats']
2023-03-16 15:36:19 [scrapy.middleware] INFO: Enabled spider middlewares:
['scrapy.spidermiddlewares.httperror.HttpErrorMiddleware',
 'scrapy.spidermiddlewares.offsite.OffsiteMiddleware',
 'scrapy.spidermiddlewares.referer.RefererMiddleware',
 'scrapy.spidermiddlewares.urllength.UrlLengthMiddleware',
 'scrapy.spidermiddlewares.depth.DepthMiddleware']
2023-03-16 15:36:19 [scrapy.middleware] INFO: Enabled item pipelines:
[]
2023-03-16 15:36:19 [scrapy.core.engine] INFO: Spider opened
2023-03-16 15:36:19 [scrapy.extensions.logstats] INFO: {
    "delta": {
        "log_count/INFO": 7,
        "log_count/WARNING": 1
    },
    "stats": {
        "log_count/INFO": 7,
        "log_count/WARNING": 1,
        "start_time": "2023-03-16 13:36:19"
    },
    "time": {
        "elapsed": 0.0,
        "log_interval": 60.0,
        "log_interval_real": 0.0,
        "start_time": "2023-03-16 13:36:19",
        "utcnow": "2023-03-16 13:36:19"
    }
}
2023-03-16 15:36:19 [scrapy.extensions.telnet] INFO: Telnet console listening on 127.0.0.1:6023
2023-03-16 15:37:19 [scrapy.extensions.logstats] INFO: {
    "delta": {
        "downloader/request_bytes": 56047,
        "downloader/request_count": 163,
        "downloader/request_method_count/GET": 163,
        "downloader/response_bytes": 582710,
        "downloader/response_count": 147,
        "downloader/response_status_count/200": 147,
        "item_scraped_count": 139,
        "log_count/INFO": 2,
        "log_count/WARNING": 0
    },
    "stats": {
        "downloader/request_bytes": 56047,
        "downloader/request_count": 163,
        "downloader/request_method_count/GET": 163,
        "downloader/response_bytes": 582710,
        "downloader/response_count": 147,
        "downloader/response_status_count/200": 147,
        "dupefilter/filtered": 160,
        "httpcompression/response_bytes": 3097674,
        "httpcompression/response_count": 147,
        "item_scraped_count": 139,
        "log_count/INFO": 9,
        "log_count/WARNING": 1,
        "request_depth_max": 8,
        "response_received_count": 147,
        "scheduler/dequeued": 163,
        "scheduler/dequeued/memory": 163,
        "scheduler/enqueued": 169,
        "scheduler/enqueued/memory": 169,
        "start_time": "2023-03-16 13:36:19"
    },
    "time": {
        "elapsed": 60.009678,
        "log_interval": 60.0,
        "log_interval_real": 60.009678,
        "start_time": "2023-03-16 13:36:19",
        "utcnow": "2023-03-16 13:37:19"
    }
}
2023-03-16 15:38:19 [scrapy.extensions.logstats] INFO: {
    "delta": {
        "downloader/request_bytes": 57356,
        "downloader/request_count": 168,
        "downloader/request_method_count/GET": 168,
        "downloader/response_bytes": 663404,
        "downloader/response_count": 168,
        "downloader/response_status_count/200": 168,
        "item_scraped_count": 158,
        "log_count/INFO": 1,
        "log_count/WARNING": 0
    },
    "stats": {
        "downloader/request_bytes": 113403,
        "downloader/request_count": 331,
        "downloader/request_method_count/GET": 331,
        "downloader/response_bytes": 1246114,
        "downloader/response_count": 315,
        "downloader/response_status_count/200": 315,
        "dupefilter/filtered": 360,
        "httpcompression/response_bytes": 6666607,
        "httpcompression/response_count": 315,
        "item_scraped_count": 297,
        "log_count/INFO": 10,
        "log_count/WARNING": 1,
        "request_depth_max": 18,
        "response_received_count": 315,
        "scheduler/dequeued": 331,
        "scheduler/dequeued/memory": 331,
        "scheduler/enqueued": 379,
        "scheduler/enqueued/memory": 379,
        "start_time": "2023-03-16 13:36:19"
    },
    "time": {
        "elapsed": 120.014717,
        "log_interval": 60.0,
        "log_interval_real": 60.005039,
        "start_time": "2023-03-16 13:36:19",
        "utcnow": "2023-03-16 13:38:19"
    }
}
2023-03-16 15:39:19 [scrapy.extensions.logstats] INFO: {
    "delta": {
        "downloader/request_bytes": 55780,
        "downloader/request_count": 162,
        "downloader/request_method_count/GET": 162,
        "downloader/response_bytes": 656491,
        "downloader/response_count": 162,
        "downloader/response_status_count/200": 162,
        "item_scraped_count": 152,
        "log_count/INFO": 1,
        "log_count/WARNING": 0
    },
    "stats": {
        "downloader/request_bytes": 169183,
        "downloader/request_count": 493,
        "downloader/request_method_count/GET": 493,
        "downloader/response_bytes": 1902605,
        "downloader/response_count": 477,
        "downloader/response_status_count/200": 477,
        "dupefilter/filtered": 540,
        "httpcompression/response_bytes": 10158375,
        "httpcompression/response_count": 477,
        "item_scraped_count": 449,
        "log_count/INFO": 11,
        "log_count/WARNING": 1,
        "request_depth_max": 27,
        "response_received_count": 477,
        "scheduler/dequeued": 493,
        "scheduler/dequeued/memory": 493,
        "scheduler/enqueued": 568,
        "scheduler/enqueued/memory": 568,
        "start_time": "2023-03-16 13:36:19"
    },
    "time": {
        "elapsed": 180.007027,
        "log_interval": 60.0,
        "log_interval_real": 59.99231,
        "start_time": "2023-03-16 13:36:19",
        "utcnow": "2023-03-16 13:39:19"
    }
}
2023-03-16 15:40:19 [scrapy.extensions.logstats] INFO: {
    "delta": {
        "downloader/request_bytes": 54634,
        "downloader/request_count": 161,
        "downloader/request_method_count/GET": 161,
        "downloader/response_bytes": 628642,
        "downloader/response_count": 161,
        "downloader/response_status_count/200": 161,
        "item_scraped_count": 153,
        "log_count/INFO": 1,
        "log_count/WARNING": 0
    },
    "stats": {
        "downloader/request_bytes": 223817,
        "downloader/request_count": 654,
        "downloader/request_method_count/GET": 654,
        "downloader/response_bytes": 2531247,
        "downloader/response_count": 638,
        "downloader/response_status_count/200": 638,
        "dupefilter/filtered": 720,
        "httpcompression/response_bytes": 13540179,
        "httpcompression/response_count": 638,
        "item_scraped_count": 602,
        "log_count/INFO": 12,
        "log_count/WARNING": 1,
        "request_depth_max": 36,
        "response_received_count": 638,
        "scheduler/dequeued": 654,
        "scheduler/dequeued/memory": 654,
        "scheduler/enqueued": 757,
        "scheduler/enqueued/memory": 757,
        "start_time": "2023-03-16 13:36:19"
    },
    "time": {
        "elapsed": 240.009861,
        "log_interval": 60.0,
        "log_interval_real": 60.002834,
        "start_time": "2023-03-16 13:36:19",
        "utcnow": "2023-03-16 13:40:19"
    }
}
2023-03-16 15:41:19 [scrapy.extensions.logstats] INFO: {
    "delta": {
        "downloader/request_bytes": 53430,
        "downloader/request_count": 158,
        "downloader/request_method_count/GET": 158,
        "downloader/response_bytes": 617247,
        "downloader/response_count": 158,
        "downloader/response_status_count/200": 158,
        "item_scraped_count": 148,
        "log_count/INFO": 1,
        "log_count/WARNING": 0
    },
    "stats": {
        "downloader/request_bytes": 277247,
        "downloader/request_count": 812,
        "downloader/request_method_count/GET": 812,
        "downloader/response_bytes": 3148494,
        "downloader/response_count": 796,
        "downloader/response_status_count/200": 796,
        "dupefilter/filtered": 920,
        "httpcompression/response_bytes": 16892151,
        "httpcompression/response_count": 796,
        "item_scraped_count": 750,
        "log_count/INFO": 13,
        "log_count/WARNING": 1,
        "request_depth_max": 46,
        "response_received_count": 796,
        "scheduler/dequeued": 812,
        "scheduler/dequeued/memory": 812,
        "scheduler/enqueued": 967,
        "scheduler/enqueued/memory": 967,
        "start_time": "2023-03-16 13:36:19"
    },
    "time": {
        "elapsed": 300.000058,
        "log_interval": 60.0,
        "log_interval_real": 59.990197,
        "start_time": "2023-03-16 13:36:19",
        "utcnow": "2023-03-16 13:41:19"
    }
}
2023-03-16 15:42:19 [scrapy.extensions.logstats] INFO: {
    "delta": {
        "downloader/request_bytes": 56378,
        "downloader/request_count": 166,
        "downloader/request_method_count/GET": 166,
        "downloader/response_bytes": 633870,
        "downloader/response_count": 166,
        "downloader/response_status_count/200": 166,
        "item_scraped_count": 162,
        "log_count/INFO": 1,
        "log_count/WARNING": 0
    },
    "stats": {
        "downloader/request_bytes": 333625,
        "downloader/request_count": 978,
        "downloader/request_method_count/GET": 978,
        "downloader/response_bytes": 3782364,
        "downloader/response_count": 962,
        "downloader/response_status_count/200": 962,
        "dupefilter/filtered": 1000,
        "httpcompression/response_bytes": 20193429,
        "httpcompression/response_count": 962,
        "item_scraped_count": 912,
        "log_count/INFO": 14,
        "log_count/WARNING": 1,
        "request_depth_max": 50,
        "response_received_count": 962,
        "scheduler/dequeued": 978,
        "scheduler/dequeued/memory": 978,
        "scheduler/enqueued": 1050,
        "scheduler/enqueued/memory": 1050,
        "start_time": "2023-03-16 13:36:19"
    },
    "time": {
        "elapsed": 360.012588,
        "log_interval": 60.0,
        "log_interval_real": 60.01253,
        "start_time": "2023-03-16 13:36:19",
        "utcnow": "2023-03-16 13:42:19"
    }
}
2023-03-16 15:42:52 [scrapy.core.engine] INFO: Closing spider (finished)
2023-03-16 15:42:52 [scrapy.extensions.logstats] INFO: {
    "delta": {
        "downloader/request_bytes": 24970,
        "downloader/request_count": 72,
        "downloader/request_method_count/GET": 72,
        "downloader/response_bytes": 344146,
        "downloader/response_count": 88,
        "downloader/response_status_count/200": 88,
        "item_scraped_count": 88,
        "log_count/INFO": 2,
        "log_count/WARNING": 0
    },
    "stats": {
        "downloader/request_bytes": 358595,
        "downloader/request_count": 1050,
        "downloader/request_method_count/GET": 1050,
        "downloader/response_bytes": 4126510,
        "downloader/response_count": 1050,
        "downloader/response_status_count/200": 1050,
        "dupefilter/filtered": 1000,
        "elapsed_time_seconds": 393.290259,
        "finish_reason": "finished",
        "finish_time": "2023-03-16 13:42:52",
        "httpcompression/response_bytes": 21921006,
        "httpcompression/response_count": 1050,
        "item_scraped_count": 1000,
        "log_count/INFO": 16,
        "log_count/WARNING": 1,
        "request_depth_max": 50,
        "response_received_count": 1050,
        "scheduler/dequeued": 1050,
        "scheduler/dequeued/memory": 1050,
        "scheduler/enqueued": 1050,
        "scheduler/enqueued/memory": 1050,
        "start_time": "2023-03-16 13:36:19"
    },
    "time": {
        "elapsed": 393.290259,
        "log_interval": 60.0,
        "log_interval_real": 33.277671,
        "start_time": "2023-03-16 13:36:19",
        "utcnow": "2023-03-16 13:42:52"
    }
}
2023-03-16 15:42:52 [scrapy.statscollectors] INFO: Dumping Scrapy stats:
{'downloader/request_bytes': 358595,
 'downloader/request_count': 1050,
 'downloader/request_method_count/GET': 1050,
 'downloader/response_bytes': 4126510,
 'downloader/response_count': 1050,
 'downloader/response_status_count/200': 1050,
 'dupefilter/filtered': 1000,
 'elapsed_time_seconds': 393.290259,
 'finish_reason': 'finished',
 'finish_time': datetime.datetime(2023, 3, 16, 13, 42, 52, 476974),
 'httpcompression/response_bytes': 21921006,
 'httpcompression/response_count': 1050,
 'item_scraped_count': 1000,
 'log_count/INFO': 17,
 'log_count/WARNING': 1,
 'request_depth_max': 50,
 'response_received_count': 1050,
 'scheduler/dequeued': 1050,
 'scheduler/dequeued/memory': 1050,
 'scheduler/enqueued': 1050,
 'scheduler/enqueued/memory': 1050,
 'start_time': datetime.datetime(2023, 3, 16, 13, 36, 19, 186715)}
2023-03-16 15:42:52 [scrapy.core.engine] INFO: Spider closed (finished)

Process finished with exit code 0


```
</details>

Gallaecio commented on scrapy/extensions/logstats.py:
```diff
@@ -1,58 +1,134 @@
     @classmethod
     def from_crawler(cls, crawler):
         interval = crawler.settings.getfloat("LOGSTATS_INTERVAL")
+        extended = crawler.settings.getbool("LOGSTATS_EXTENDED_ENABLED")
```
Assuming we may eventually get a non-delta alternative implementation to address https://github.com/scrapy/scrapy/issues/2173, I think it may be worth considering to replace the boolean `LOGSTATS_EXTENDED_ENABLED` with a string `LOGSTATS_MODE` that supports `"simple"` (default) and `"delta"`, and in the future also `"stats"`.

We can then remove `_EXTENDED` from the names of the other 2 new settings.

Gallaecio reviewed (commented):
OK, I think we should stop reviewing this as a solution for https://github.com/scrapy/scrapy/issues/2173, and review this as a proposal to log stat deltas.

And, if we do that, I think we need to go back to a simpler output that only shows delta. This approach of logging separate groups of data (delta, time, stats) feels unnecessarily complex to me.

I would like going back to the [original output](https://github.com/scrapy/scrapy/pull/5830#issue-1577976674), only changing the messaging to something more clear like “Stat changes in the last 60.0 seconds:”.

I would also suggest providing some sane default for the include and exclude settings. Hopefully a value that would also make sense for a future “stats mode” that logs raw stats rather than deltas.

GeorgeA92 commented:
> And, if we do that, I think we need to go back to a simpler output that only shows delta. This approach of logging separate groups of data (delta, time, stats) feels unnecessarily complex to me.

1. stats - needed as it is directly requested in https://github.com/scrapy/scrapy/issues/2173
2. deltas - also needed. If application already have implementation of periodial log stats - sooner or later we will inevitably receive request for.. per-minute changes of stats as obvious consequence of this. As this is technically the most  - I initially focused to implement this at first.
3. timing data - also needed due to reasons I mentioned earlier.

And I agree that on including all 3 parts of data.. looks too much for existing built-in extension.
As we already have implementation It is preferable to include all of this in scrapy (maybe in another form) .. Separate extension or/and separate log entry per each segment of stats data
